nequip/data/_dataset/_hdf5_dataset.py

```diff
@@ -0,0 +1,133 @@
+                        self.index.append((types, pos, energy, forces, i))
+                else:
+                    for i in range(len(energy)):
+                        self.index.append((types, pos, energy, i))
+
+    def len(self) -> int:
+        return self.num_molecules
+
+    def get(self, idx: int) -> AtomicData:
+        if self.index is None:
+            self.setup_index()
+        data = self.index[idx]
+        i = data[-1]
+        args = {
+            "pos": data[1][i],
+            "r_max": self.r_max,
+            AtomicDataDict.ATOM_TYPE_KEY: data[0][i],
+            AtomicDataDict.TOTAL_ENERGY_KEY: data[2][i],
+        }
+        if self.has_forces:
+            args[AtomicDataDict.FORCE_KEY] = data[3][i]
+        return AtomicData.from_points(**args)
+
+    def statistics(
+        self,
+        fields: List[Union[str, Callable]],
+        modes: List[str],
+        stride: int = 1,
+        unbiased: bool = True,
+        kwargs: Optional[Dict[str, dict]] = {},
+    ) -> List[tuple]:
+        assert len(modes) == len(fields)
+        # TODO: use RunningStats
+        if len(fields) == 0:
+            return []
+        if self.index is None:
+            self.setup_index()
+        results = []
+        for field, mode in zip(fields, modes):
+            count = 0
+            if mode == "rms":
+                total = 0.0
+            elif mode in ("mean_std", "per_atom_mean_std"):
+                total = [0.0, 0.0]
+            else:
+                raise NotImplementedError(f"Analysis mode '{mode}' is not implemented")
+            for index in range(0, self.len(), stride):
+                data = self.index[index]
+                i = data[-1]
+                if field == AtomicDataDict.FORCE_KEY:
+                    values = data[3][i]
+                elif field == AtomicDataDict.TOTAL_ENERGY_KEY:
+                    values = data[2][i]
+                elif callable(field):
+                    values, _ = field(self.get(index))
+                    values = np.asarray(values)
+                else:
+                    raise RuntimeError(
+                        f"The field key `{field}` is not present in this dataset"
+                    )
+                if mode == "rms":
+                    total += np.sum(values * values)
+                    count += len(values.flatten())
+                else:
+                    length = len(values.flatten())
+                    if mode == "per_atom_mean_std":
+                        values /= len(data[0][i])
+                    sample_mean = np.mean(values)
+                    new_mean = (total[0] * count + sample_mean * length) / (
+                        count + length
+                    )
+                    total[1] += (
+                        length * (sample_mean - total[0]) * (sample_mean - new_mean)
+                    )
+                    total[0] = new_mean
+                    count += length
```

Comment:
See https://github.com/mir-group/pytorch_runstats / https://pytorch-runstats.readthedocs.io/en/latest/index.html  

which can do this for `rms` and `mean` already with unit tests and correct numerics.  We already depend on this library.

Easy to do an OK std with two-pass using this (https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Two-pass_algorithm) or to implement Welford's online algorithm into `torch_runstats` maybe (https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm)?

Or can just return `NaN` for the std if you only ever use the mean for now.